### PR TITLE
Compatability/conv belts

### LINKED ISSUE #2196: Lumberjacks want to collect dropped items from Conveyor Belts
I jave a lumberjack that has trees to cut, but near the farm I have a Conveyor Belt system (from Immersive Engineering). The items on conveyor belts keep moving as usual, but the NPC always want to collect the items from the conveyors, and don't cut any tree, as he's on a loop trying to get into the conveyor belts.

NPC info:
```
> [17:07:24] [Server thread/INFO]: ID: 7 Name: Oliver Y. Ashcombe
> [17:07:24] [Server thread/INFO]: Citizen position: x=-22336 y=63 z=8757
> [17:07:24] [Server thread/INFO]: Home position: x=-22372 y=71 z=8777
> [17:07:24] [Server thread/INFO]: Work position: x=-22332 y=63 z=8757
> [17:07:24] [Server thread/INFO]: Health: 20.0 Max Health: 20.0
> [17:07:24] [Server thread/INFO]: Level: 0 Age: 0 Experience: 0
> [17:07:24] [Server thread/INFO]: Charisma: 6 Dexterity: 2 Endurance: 7
Intelligence: 2 Strength: 3
> [17:07:24] [Server thread/INFO]: Job: Lumberjack
> [17:07:24] [Server thread/INFO]: Desired activity: WORK Current Activity: LUMBERJACK_GATHERING
```

EDIT: I think this is going to need a hardcoded check. Immersive Engineering spawn and moves the items on a vanilla way (spawning the EntityItem and changing it's motion), so there's no clean way to detect if an item is being carried by a conveyor belt. An *ugly* fix that I can think of is blacklisting NPCS from picking up items that are in the same position as a TileEntity (aka inside the TE, if the TE holds an item I don't think you're supposed to pick it up), or add an hardcoded check to see if there's a Conveyor Belt near the item.

### LINKED ISSUE #2251: 1.12.2 crash when clicking upgrade building on citizen hut level 4
### Minecolonies version

version 0.8.7613

### Expected behavior

not to crash when upgrading building

### Actual behavior

client crashes when clicking upgrade on citizen hut 4

crash report: https://gist.github.com/Faenum/d21b2a5163d2483206d0f098b38ceb2a

### LINKED ISSUE #2252: NullPointerException: With Citizens trying to do thier jobs. (AiTasks) (not schematic related)
### Version Info
**Forge** `v14.23.2.2611`
**Minecolonies** `1.12.2-0.8.7613`

### Expected behaviour

- _I actually do not know if this is normal or not_, this error has happened only after installed the latest alpha however it pauses the baker work order.

### Actual behaviour
`Action for target com.minecolonies.coremod.entity.ai.util.AITarget@4aca7725 threw an exception` -
`[23:56:19] [Server thread/ERROR] [minecolonies]: Pausing Entity Hugo J. Abell (com.minecolonies.coremod.job.baker) at BlockPos{x=221, y=67, z=387} for 100 Seconds because of error:
[23:56:23] [Server thread/WARN] [minecolonies]: Action for target com.minecolonies.coremod.entity.ai.util.AITarget@4aca7725 threw an exception:`
- [Pastebin](https://pastebin.com/pBwabMvy)
- [Gitgist](https://gist.github.com/Misunderstood-Wookiee/ac853e6973ef51a1d72ec67cca99766e)

### Steps to reproduce the problem
Not sure, sorry. I was not playing at the time, it happened on the server.

### LINKED ISSUE #2261: [Security Feature] Check for overworld placement of TH, Supply Ship/Camp.
Make a "check" function to determine if the Supply Ship/Camp, Town Hall are being placed in the "Overworld". 
If not then deny placement and sent message: "You can only place this in the Overworld dimension".
Also have an option to turn this check ON/OFF in Config (for future use on other dimensions and server owners control).

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #2252
Closes #2251
Closes #2196
Closes #2261

# Changes proposed in this pull request:
- Fix npe
- Fix schematics
- Fix convoyer belts
- Make waypoints blocks passable
- Make substitution blocks have their texture on all sides
- Adds a config value to avoid colonies placed outside of the overworld.

Review please

### CODE REVIEW

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 58 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 49 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 4 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [AbstractBuildingWorker.java#L88](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L88): Remove this call from a constructor to the overridable "getBuildingLevel" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L583](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L583): Reduce this class from 84 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/c045bfa41c91a92f0864115efa909ecf415f6c63/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 1]
```diff
@@ -291,7 +291,8 @@ public void searchForItems(final AxisAlignedBB boundingBox)
     {
         items = world.getEntitiesWithinAABB(EntityItem.class, boundingBox)
                   .stream()
-                  .filter(item -> item != null && !item.isDead)
+                  .filter(item -> item != null && !item.isDead &&
+                          (!item.getEntityData().hasKey("PreventRemoteMovement") || !item.getEntityData().getBoolean("PreventRemoteMovement")))
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java:295]: Shouldn't `!item.getEntityData().getBoolean("PreventRemoteMovement")` be `item.getEntityData().getBoolean("PreventRemoteMovement")`?

**Review by marchermans — COMMENTED** [review]
Small questions with regards to conveyor belts.

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java (hunk 1 above)
Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java:295]: or the `||` be an end `&&`?

Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java:295]: Never mind.

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java:295]: Great that this is settled =)

**Review by marchermans — DISMISSED** [review]

File: src/main/java/com/minecolonies/coremod/colony/LoadColonyThread.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                @NotNull final ChunkLoadStorage newStorage = new ChunkLoadStorage(id, ChunkPos.asLong(i, j), add, dimension, owning);
+                @NotNull final File file = new File(chunkDir, String.format(FILENAME_CHUNK, i, j, dimension));
+                if (file.exists())
+                {
+                    @Nullable final NBTTagCompound chunkData = loadNBTFromPath(file);
+                    final ChunkLoadStorage storage = new ChunkLoadStorage(chunkData);
+                    storage.merge(newStorage);
+                    if (storage.isEmpty())
+                    {
+                        file.delete();
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/LoadColonyThread.java]: MINOR Do something with the "boolean" value returned by "delete". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS899)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 3]
```diff
@@ -229,46 +229,11 @@ public static void notifyChunksInRange(final World world, final boolean add, fin
...
-                {
-                    saveNBTToPath(file, newStorage.toNBT());
-                    missingChunksToLoad++;
-                }
-            }
-        }
+        final LoadColonyThread thread = new LoadColonyThread(centralChunk.x, centralChunk.z, id, dimension, add);
+        thread.start();
+        missingChunksToLoad = thread.getMissingChunksToLoad();
+        //todo store thread in ColonyManager and get value each time we need "missingChunksToLoad", if thread is null, then no need to get it
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 4]
```diff
@@ -229,46 +229,11 @@ public static void notifyChunksInRange(final World world, final boolean add, fin
...
+        //todo on shut down do thread.join and then get missingChunksToLoad
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 78 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 59 major
* ![MINOR][MINOR] 11 minor
* ![INFO][INFO] 6 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [ColonyManager.java#L915](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L915): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![CRITICAL][CRITICAL] [AbstractBuildingWorker.java#L88](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L88): Remove this call from a constructor to the overridable "getBuildingLevel" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [ColonyManager.java#L55](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L55): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 39 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [ColonyManager.java#L55](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L55): class "ColonyManager" has 57 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [ColonyManager.java#L296](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L296): Provide multiple methods instead of using "canDestroy" to determine which action to take. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2301)
1. ![MAJOR][MAJOR] [ColonyManager.java#L561](https://github.com/Minecolonies/minecolonies/blob/418e80514e01180c60cabc0f30b1fcc722333e08/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L561): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 63 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 51 major
* ![MINOR][MINOR] 7 minor
* ![INFO][INFO] 4 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [AbstractBuildingWorker.java#L88](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L88): Remove this call from a constructor to the overridable "getBuildingLevel" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L583](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L583): Reduce this class from 84 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [requestable#](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/colony/requestable#): Add a 'package-info.java' file to document the 'requestable' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [data#](https://github.com/Minecolonies/minecolonies/blob/3525a8eb6a7cb3d980b77b74a11a2b7d44748987/src/main/java/com/minecolonies/coremod/colony/requestsystem/data#): Add a 'package-info.java' file to document the 'data' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — APPROVED** [review]

**Comment by Misunderstood-Wookiee** [comment]
The conveyor belts fix, this is for the engineering mod? To fix loggers taking logs from the belts? Was waiting for a fix before installation of that mod as I was going to make schematics with the water Mills :)

**Comment by Raycoms** [comment]
yes

**Comment by Misunderstood-Wookiee** [comment]
Oh nice!

**Pull request merged**